yew-macro/src/html_tree/html_if.rs

```diff
@@ -0,0 +1,128 @@
+use super::HtmlRoot;
+use crate::PeekValue;
+use boolinator::Boolinator;
+use proc_macro2::{Delimiter, TokenStream};
+use quote::{quote, quote_spanned, ToTokens};
+use syn::buffer::Cursor;
+use syn::parse::{Parse, ParseStream, Result as ParseResult};
+use syn::spanned::Spanned;
+use syn::{braced, token, Expr, Token};
+
+pub struct HtmlIf {
+    if_token: Token![if],
+    cond: Box<Expr>,
+    then_branch: HtmlBranch,
+    else_branch: Option<(Token![else], Box<HtmlBranchOrIf>)>,
+}
+
+impl PeekValue<()> for HtmlIf {
+    fn peek(cursor: Cursor) -> Option<()> {
+        let (ident, _) = cursor.ident()?;
+        (ident == "if").as_option()
+    }
+}
+
+impl Parse for HtmlIf {
+    fn parse(input: ParseStream) -> ParseResult<Self> {
+        let if_token = input.parse()?;
+        let cond = input.parse()?;
+        let then_branch = input.parse()?;
+        let else_branch = input
+            .parse::<Token![else]>()
+            .ok()
+            .map(|else_token| input.parse().map(|branch| (else_token, branch)))
+            .transpose()?;
+
+        Ok(HtmlIf {
+            if_token,
+            cond,
+            then_branch,
+            else_branch,
+        })
+    }
+}
+
+impl ToTokens for HtmlIf {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let HtmlIf {
+            if_token,
+            cond,
+            then_branch,
+            else_branch,
+        } = self;
+        let default_else_branch = syn::parse_str("{}").unwrap();
+        let else_branch = else_branch
+            .as_ref()
+            .map(|(_, branch)| branch)
+            .unwrap_or(&default_else_branch);
+        let new_tokens = quote_spanned! {if_token.span()=>
+            if #cond #then_branch else #else_branch
+        };
+
+        tokens.extend(new_tokens);
+    }
+}
+
+pub struct HtmlBranch {
+    _brace: token::Brace,
+    root: HtmlRoot,
```

Comment:
I just realized this makes it impossible to have multiple nodes in a branch:
```rust
html! {
    <div>
        if true {
            <span>{ "Hello" }</span>
            <span>{ "World" }</span>
            // doesn't work :(
        }
    </div>
}
```

It would be great if we supported that though, because using fragments isn't ideal (and doesn't work when dealing with components with typed children).

Not quite sure how to best approach this though. One thing is certain, with this change we can no longer allow `HtmlIf` at the root.

If you don't have an idea how to implement it we can just postpone it.

Reply:
I was thinking about using `HtmlChildrenTree`.

I did make an attempt but it failed... I will try again

Reply:
Please have a look 9d6af31 :eyes: